testsuite/model/src/test/java/org/keycloak/testsuite/model/session/AuthenticationSessionTest.java

```diff
@@ -198,4 +200,58 @@ public void testRemoveExpiredAuthSessions() {
             return null;
         });
     }
+
+    @Test
+    public void testConcurrentAuthenticationSessionsCreation() throws InterruptedException {
+        final String rootId = withRealm(realmId, (session, realm) -> {
+            RootAuthenticationSessionModel rootAuthSession = session.authenticationSessions().createRootAuthenticationSession(realm);
+            return rootAuthSession.getId();
+        });
+        ConcurrentHashMap.KeySetView<String, Boolean> tabIds = ConcurrentHashMap.newKeySet();
+        inIndependentFactories(4, 60, () -> {
+                withRealm(realmId, (session, realm) -> {
+                    RootAuthenticationSessionModel rootAuthSession = session.authenticationSessions().getRootAuthenticationSession(realm, rootId);
+                    ClientModel client = realm.getClientByClientId("test-app");
+                    AuthenticationSessionModel authenticationSession = rootAuthSession.createAuthenticationSession(client);
+                    tabIds.add(authenticationSession.getTabId());
+                    return null;
+                });
+        });
+
+        withRealm(realmId, (session, realm) -> {
+            RootAuthenticationSessionModel rootAuthSession = session.authenticationSessions().getRootAuthenticationSession(realm, rootId);
+            Assert.assertEquals(4, rootAuthSession.getAuthenticationSessions().size());
+            assertThat(rootAuthSession.getAuthenticationSessions().keySet(), Matchers.containsInAnyOrder(tabIds.toArray()));
+            return null;
+        });
+    }
+
+    @Test
+    public void testConcurrentAuthenticationSessionsRemoval() throws InterruptedException {
+        ConcurrentLinkedQueue<String> tabIds = new ConcurrentLinkedQueue<>();
+        int concurrentTabs = 4;
+        final String rootId = withRealm(realmId, (session, realm) -> {
+            RootAuthenticationSessionModel rootAuthSession = session.authenticationSessions().createRootAuthenticationSession(realm);
+            ClientModel client = realm.getClientByClientId("test-app");
+
+            for (int i = 0; i < concurrentTabs; i++) {
+                tabIds.add(rootAuthSession.createAuthenticationSession(client).getTabId());
+            }
+            return rootAuthSession.getId();
+        });
+        inIndependentFactories(concurrentTabs, 60, () -> {
+            withRealm(realmId, (session, realm) -> {
+                RootAuthenticationSessionModel rootAuthSession = session.authenticationSessions().getRootAuthenticationSession(realm, rootId);
+                rootAuthSession.removeAuthenticationSessionByTabId(tabIds.remove());
+                return null;
+            });
+        });
+
+        withRealm(realmId, (session, realm) -> {
+            RootAuthenticationSessionModel rootAuthSession = session.authenticationSessions().getRootAuthenticationSession(realm, rootId);
+            //TODO: rootAuthSession should be null not just empty check removeAuthenticationSessionByTabId
```

Comment:
@pruivo We need to figure out how to remove the root auth session when there are no more auth sessions. The check should happen after the update tasks are finished. See `removeAuthenticationSessionByTabId` method